### PR TITLE
When mention `@channel/@all` show the timezones affected in the warning message

### PR DESCRIPTION
**Pull request by cpanato**
#### Summary
screen shot 2018-08-23 at 13 55 37

#### Ticket Link
n/a

#### Other Prs:
server: https://github.com/mattermost/mattermost-server/pull/9286
redux: https://github.com/mattermost/mattermost-redux/pull/612
api reference: https://github.com/mattermost/mattermost-api-reference/pull/387

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)
- [x] Has server changes (please link)
- [x] Has redux changes (please link)
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-webapp/blob/master/i18n/en.json)) updates

### CODE REVIEW

**Comment by esethna** [comment]
@cpanato is there any discussion on the feature on pre-release? Is this only shown if the experimental timezone setting is enabled? My concerns are:

1. Don't see much value in having the timezones written in the dialog as the purpose of that dialog is to inform the user that they are going to be sending a lot of notifications. If they are aware of that fact, I don't think knowing the number of timezones would change my decision to send a mention or not. It might in a DM, but not for a channel wide mention. 
2. The message itself is not super helpful in my mind. ie if I'm sending a message to channel with only canadians in it, it still might say "4 different timezones" even though they are only a few hours apart. But I could get the same message if someone from africa or england was in the channel, but the time zones are much farther apart.
3. Does it base it on the timezone of the latest client used to access Mattermost by any users in the channel. If it's based on the experimental setting some users will not have a timezone set (ie using "automatic".
4. Do you have any references to other software doing this? I don't think I've seen this in other products

**Comment by esethna** [comment]
@cpanato so this is only shown right now if the experimental timezone feature is enabled?

**Comment by cpanato** [comment]
@esethna yes, it will only show if the experimental timezone is enabled. when we remove the experimental than we can show always

**Comment by esethna** [comment]
@cpanato can you help enable the timezone feature on this test server when you have a chance?

**Comment by cpanato** [comment]
@esethna I guess this will not work because it depends of the server and redux changes. not sure how we can test it here.

but I've enable the timezone

**Comment by cpanato** [comment]
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

**Comment by cpanato** [comment]
Spintmint test server creation failed. See the logs for more information.

**Comment by esethna** [comment]
Hey @cpanato, please let me know when it's ready to test. Looks like redux changes are still waiting to be merged?

**Comment by cpanato** [comment]
@esethna test server ready and the timezone setting enabled.

**Comment by cpanato** [comment]
@esethna done

**Comment by esethna** [comment]
Thanks, please help enable the timezone setting ^

**Comment by cpanato** [comment]
@esethna done

**Comment by esethna** [comment]
@cpanato, one more tweak, looks like we are pluralizing the word "timezone" even if there's just one:

image


Should just be "timezone"

**Comment by cpanato** [comment]
@esethna done the plural

**Comment by esethna** [comment]
Thanks @cpanato, please help enable the timezone feature ^

**Comment by cpanato** [comment]
@esethna done

File: components/create_comment/create_comment.jsx [hunk 1]
```diff
@@ -167,6 +173,11 @@ export default class CreateComment extends React.PureComponent {
          */
         maxPostSize: PropTypes.number.isRequired,
         rhsExpanded: PropTypes.bool.isRequired,
+
+        /**
+         * To check if the timezones are enable in the server.
```
[the remark this answers is not in the document]
Comment (cpanato) [reply components/create_comment/create_comment.jsx]: done

File: components/create_comment/create_comment.jsx [hunk 2]
```diff
@@ -167,6 +173,11 @@ export default class CreateComment extends React.PureComponent {
...
+         */
+        isTimezoneEnable: PropTypes.bool.isRequired,
```
[the remark this answers is not in the document]
Comment (cpanato) [reply components/create_comment/create_comment.jsx]: done

File: components/create_comment/create_comment.jsx [hunk 3]
```diff
@@ -181,6 +192,7 @@ export default class CreateComment extends React.PureComponent {
                 uploadsInProgress: [],
                 fileInfos: [],
             },
+            channelTimezones: -1,
```
[the remark this answers is not in the document]
Comment (cpanato) [reply components/create_comment/create_comment.jsx]: I like to be magical 

done

File: components/create_post/create_post.jsx [hunk 4]
```diff
@@ -875,15 +899,29 @@ export default class CreatePost extends React.Component {
...
-            <FormattedMessage
-                id='notify_all.question'
-                defaultMessage='By using @all or @channel you are about to send notifications to {totalMembers} people. Are you sure you want to do this?'
-                values={{
-                    totalMembers: members,
-                }}
-            />
-        );
+        let notifyAllMessage = '';
+        if (this.state.channelTimezones !== -1 && this.props.isTimezoneEnable) {
```
[the remark this answers is not in the document]
Comment (cpanato) [reply components/create_post/create_post.jsx]: done

File: components/create_comment/create_comment.jsx [hunk 5]
```diff
@@ -337,6 +349,18 @@ export default class CreateComment extends React.PureComponent {
     handleSubmit = async (e) => {
         e.preventDefault();
 
+        if (this.props.isTimezoneEnable) {
+            this.props.getChannelTimezones(this.props.channelId).then(
+                (data) => {
+                    if (data.data) {
+                        this.setState({channelTimezones: data.data.length});
```
[the remark this answers is not in the document]
Comment (cpanato) [reply components/create_comment/create_comment.jsx]: done

[the remark this answers is not in the document]
Comment (cpanato) [reply components/create_comment/create_comment.jsx]: thanks for the review

File: components/create_post/create_post.jsx [hunk 6]
```diff
@@ -395,6 +407,18 @@ export default class CreatePost extends React.Component {
             userIsOutOfOffice,
         } = this.props;
 
+        if (this.props.isTimezoneEnabled) {
+            this.props.actions.getChannelTimezones(this.props.currentChannel.id).then(
```
Comment (jwilander) [inline components/create_post/create_post.jsx]: No action needed just a passing comment: I was going to suggest switching this to use `await` but that would delay the modal opening on the network request so I think this makes more sense

File: components/create_comment/create_comment.jsx [hunk 7]
```diff
@@ -652,15 +676,29 @@ export default class CreateComment extends React.PureComponent {
...
-        );
+        let notifyAllMessage = '';
+        if (this.state.countChannelTimezones && this.props.isTimezoneEnabled) {
+            notifyAllMessage = (
+                <FormattedMarkdownMessage
+                    id='notify_all.question_timezone'
+                    defaultMessage='By using @all or @channel you are about to send notifications to **{totalMembers} people** in **{timezones, number} {timezones, plural, one {timezone} other {timezones}}**. Are you sure you want to do this?'
+                    values={{
+                        totalMembers: this.props.channelMembersCount - 1,
+                        timezones: this.state.countChannelTimezones,
```
Comment (lieut-data) [inline components/create_comment/create_comment.jsx]: Thanks for addressing the feedback, @cpanato! Just realized my suggestion for `countChannelTimezones` would have been better as `channelTimezonesCount` to match the incoming `channelMembersCount` prop. 0/5, and not blocking :)

**Pull request merged**